ScrollingStateScrollingNode destructor should be able to handle being called before being attached to a tree.

lericaa opened the pull request:
#### 69441586e9e090843c0bb55f9a8b7eb97b1521e4
<pre>
ScrollingStateScrollingNode destructor should be able to handle being called before being attached to a tree.
<a href="https://bugs.webkit.org/show_bug.cgi?id=264722">https://bugs.webkit.org/show_bug.cgi?id=264722</a>
<a href="https://rdar.apple.com/117803479">rdar://117803479</a>

Reviewed by Alex Christensen.

Adding null check to handle the case when destructor being called before being attached to a tree.

* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::isAttachedToScrollingStateTree const):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::~ScrollingStateScrollingNode):

Canonical link: <a href="https://commits.webkit.org/270696@main">https://commits.webkit.org/270696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/206f8ccf99578487c4bafe92bc06dc91bd810ce9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26113 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4723 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27386 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28209 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23910 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2136 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23954 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3579 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22478 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28785 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3189 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23436 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29487 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23807 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23820 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27377 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3231 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1427 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4636 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6288 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3697 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3552 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/scrolling/ScrollingStateNode.h:
```diff
@@ -304,6 +304,7 @@ class ScrollingStateNode : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
     const LayerRepresentation& layer() const { return m_layer; }
     WEBCORE_EXPORT void setLayer(const LayerRepresentation&);
 
+    bool attachToScrollingStateTree() const { return !!m_scrollingStateTree; }
```
How about?
```suggestion
    bool isAttachedToScrollingStateTree() const { return !!m_scrollingStateTree; }
```

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -137,7 +137,9 @@ ScrollingStateScrollingNode::ScrollingStateScrollingNode(const ScrollingStateScr
 ScrollingStateScrollingNode::~ScrollingStateScrollingNode()
 {
-    scrollingStateTree().scrollingNodeRemoved();
+    // handle the case when destructor being called before being attached to a tree
```
This comment doesn't really add much of an explanation above just reading the if statement.  Could you remove it before merging the PR?

lericaa replied:
Sure! Updated

lericaa committed:
ScrollingStateScrollingNode destructor should be able to handle being called before being attached to a tree.
https://bugs.webkit.org/show_bug.cgi?id=264722
rdar://117803479

Reviewed by Alex Christensen.

Adding null check to handle the case when destructor being called before being attached to a tree.

* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::isAttachedToScrollingStateTree const):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::~ScrollingStateScrollingNode):

Canonical link: https://commits.webkit.org/270696@main
